File: src/Adaptive.ReactiveTrader.Client.GUI/UI/SpotTiles/SpotTilesViewModel.cs
using System;
using System.Collections.ObjectModel;
using System.Linq;
using System.Reactive.Disposables;
using System.Reactive.Linq;
using Adaptive.ReactiveTrader.Client.Concurrency;
using Adaptive.ReactiveTrader.Client.Domain;
using Adaptive.ReactiveTrader.Client.Domain.Models;
using Adaptive.ReactiveTrader.Client.Domain.Models.ReferenceData;
using Adaptive.ReactiveTrader.Shared.Extensions;
using Adaptive.ReactiveTrader.Shared.UI;
using Adaptive.ReactiveTrader.Client.Domain.Repositories;
using log4net;
using PropertyChanged;

namespace Adaptive.ReactiveTrader.Client.UI.SpotTiles
{
    [ImplementPropertyChanged]
    public class SpotTilesViewModel : ViewModelBase, ISpotTilesViewModel, IDisposable
    {
        private static readonly ILog Log = LogManager.GetLogger(typeof (SpotTilesViewModel));
        
        public ObservableCollection<ISpotTileViewModel> SpotTiles { get; private set; }
        private readonly CompositeDisposable _subscriptions = new CompositeDisposable();
        private readonly IReferenceDataRepository _referenceDataRepository;
        private readonly Func<ICurrencyPair, SpotTileSubscriptionMode, ISpotTileViewModel> _spotTileFactory;
        private readonly IConcurrencyService _concurrencyService;
        private readonly ISpotTileViewModel _config;

        public SpotTilesViewModel(IReactiveTrader reactiveTrader,
            Func<ICurrencyPair, SpotTileSubscriptionMode, ISpotTileViewModel> spotTileFactory,
            IConcurrencyService concurrencyService)
        {
            _referenceDataRepository = reactiveTrader.ReferenceData;
            _spotTileFactory = spotTileFactory;
            _concurrencyService = concurrencyService;

            SpotTiles = new ObservableCollection<ISpotTileViewModel>();

            _config = spotTileFactory(null, SpotTileSubscriptionMode.Conflate);
            _config.ToConfig();

            SpotTiles.Add(_config);

            _subscriptions.Add(
                _config.Config.ObserveProperty(p => p.SubscriptionMode)
                    .Subscribe(subscriptionMode => SpotTiles.Where(vm => vm.Pricing != null).ForEach(vm => vm.Pricing.SubscriptionMode = subscriptionMode)));

            _subscriptions.Add(
                _config.Config.ObserveProperty(p => p.ExecutionMode)
                    .Subscribe(executionMode => SpotTiles.Where(vm => vm.Pricing != null).ForEach(vm => vm.Pricing.ExecutionMode = executionMode)));

            LoadSpotTiles();
        }

        public void Dispose()
        {
            _subscriptions.Dispose();
        }

        private void LoadSpotTiles()
        {
            _referenceDataRepository.GetCurrencyPairsStream()
                .ObserveOn(_concurrencyService.Dispatcher)
                .SubscribeOn(_concurrencyService.ThreadPool)
                .Subscribe(
                    currencyPairs => currencyPairs.ForEach(HandleCurrencyPairUpdate),
                    error => Log.Error("Failed to get currencies", error));
        }

        private void HandleCurrencyPairUpdate(ICurrencyPairUpdate update)
        {
            var spotTileViewModel = SpotTiles.FirstOrDefault(stvm => stvm.CurrencyPair == update.CurrencyPair.Symbol);
            if (update.UpdateType == UpdateType.Add)
            {
                if (spotTileViewModel != null)
                {
                    // we already have a tile for this currency pair
                    return;
                }

                var spotTile = _spotTileFactory(update.CurrencyPair, _config.Config.SubscriptionMode);
                SpotTiles.Add(spotTile);
            }
            else
            {
                if (spotTileViewModel != null)
                {
                    SpotTiles.Remove(spotTileViewModel);
                    spotTileViewModel.Dispose();
                }
            }
        }
    }
}

File: src/Adaptive.ReactiveTrader.Client.JS/UI/SpotTiles/SpotTilesViewModel.ts
class SpotTilesViewModel implements ISpotTilesViewModel {
    spotTiles: KnockoutObservableArray<ISpotTileViewModel>;
    private _referenceDataRepository: IReferenceDataRepository;
    private _spotTileViewModelFactory: ISpotTileViewModelFactory;

    constructor(referenceDataRepository: IReferenceDataRepository, spotTileViewModelFactory: ISpotTileViewModelFactory) {
        this._referenceDataRepository = referenceDataRepository;
        this._spotTileViewModelFactory = spotTileViewModelFactory;
        this.spotTiles = ko.observableArray([]);

        this.loadSpotTiles();
    }

    private loadSpotTiles(): void {
        this._referenceDataRepository.getCurrencyPairs()
            .subscribe(
                currencyPairs=> currencyPairs.forEach(cp=> this.handleCurrencyPairUpdate(cp)),
                ex=> console.error("Failed to get currencies", ex));
    }

    private handleCurrencyPairUpdate(update: ICurrencyPairUpdate) {
        var spotTileViewModel = ko.utils.arrayFirst(this.spotTiles(), stvm=> stvm.symbol == update.currencyPair.symbol);
        if (update.updateType == UpdateType.Add) {
            if (spotTileViewModel != null) {
                // we already have a tile for this ccy pair
                return;
            }

            var spotTile = this._spotTileViewModelFactory.create(update.currencyPair);
            this.spotTiles.push(spotTile);
        } else {
            if (spotTileViewModel != null) {
                this.spotTiles.remove(spotTileViewModel);
                spotTileViewModel.dispose();
            }                
        }
    }
} 